modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleLocalServiceImpl.java

```diff
@@ -308,6 +308,8 @@ public ObjectValidationRule updateObjectValidationRule(
 		_objectValidationRuleSettingPersistence.removeByObjectValidationRuleId(
 			objectValidationRuleId);
 
+		_objectValidationRuleSettingPersistence.flush();
+
```

Comment:
Hey @carolmariaabb ,

It is possible to fix the bug by adding the `objectValidationRuleSettings` in the `TransactionCommitCallbackUtil.registerCallback`.

But I would argue that doing this will make our code less coherent and more complex. 
Because we would be returning the `objectValidationRule` with the old settings while we intend to add the new ones after the transaction is committed.

This will also cause our current integration test to fail [here](https://github.com/liferay/liferay-portal/blob/master/modules/apps/object/object-test/src/testIntegration/java/com/liferay/object/service/test/ObjectValidationRuleLocalServiceTest.java#L324).

We could get around this by calling `objectValidationRule.setObjectValidationRuleSettings` with the new settings. But we would be returning the `objectValidationRule` with the new settings before actually adding the settings.

With this in mind, should we move on with this? Or do you see a way to avoid these drawbacks?

Reply:
@devpedrotavares we won't return the old settings and the test won't fail if we do:

```
private List<ObjectValidationRuleSetting> _addObjectValidationRuleSettings(
	ObjectValidationRule objectValidationRule,
	List<ObjectValidationRuleSetting> objectValidationRuleSettings) {

	return TransformUtil.transform(
		objectValidationRuleSettings,
		objectValidationRuleSetting ->  {
			TransactionCommitCallbackUtil.registerCallback(
				() -> {
					_objectValidationRuleSettingLocalService.
						addObjectValidationRuleSetting(
							objectValidationRule.getUserId(),
							objectValidationRule.getObjectValidationRuleId(),
							objectValidationRuleSetting.getName(),
							objectValidationRuleSetting.getValue());

					return null;
				});

			return objectValidationRuleSetting;
		});
}
```
I also don't think that this would make the code more complex.

>  But we would be returning the objectValidationRule with the new settings before actually adding the settings.

The same thing that happens in this [code](https://github.com/liferay/liferay-portal/blob/02a1313f6cf211c198351eeafcacfc05a462da48/modules/apps/segments/segments-service/src/main/java/com/liferay/segments/service/impl/SegmentsExperimentLocalServiceImpl.java#L441-L449) which returns the `SegmentsExperience` before updating its priortity.

We already know that the solution with `flush` will be rejected by Brian, as it was 3 weeks ago. 

That said, the only alternative I see to not following both approaches would be: instead of deleting all settings when updating the `ObjectValidationRule`, we could check if the new setting already exists in the database: 

1. If it doesn't exist, we can add;
2. If it exists, we don't need to do anything;

We should also make sure to delete the old settings that exist in the database but are not present in the `List<ObjectValidationRuleSetting> objectValidationRuleSettings`.

Reply:
I see yours points. I tried this last suggestion

Reply:
I understand. I tried the last suggestion.

Btw, thank you for finding that [PR](https://github.com/brianchandotcom/liferay-portal/pull/141626#issuecomment-1745001103)